File: examples/query_operators_example.py
"""
Example demonstrating ORM-style query system with Flexmodel
"""

from pymongo import MongoClient
from flexschema import Schema, Flexmodel, field


class Product(Flexmodel):
    schema: Schema = Schema.ident(
        name=field(str, nullable=False),
        price=field(float, default=0.0),
        category=field(str, nullable=False),
        in_stock=field(bool, default=True),
        quantity=field(int, default=0),
    )


if __name__ == "__main__":
    print("=== ORM-Style Query API Examples ===\n")

    try:
        client = MongoClient("mongodb://localhost:27017/testdb", serverSelectionTimeoutMS=1000)
        Product.attach(client, "products")
        select = Product.select()
        print("✓ Using MongoDB\n")
    except Exception:
        print("⚠️  MongoDB not available. Falling back to demo mode.\n")
        select = Product.select()

    print("1. Simple comparison:")
    query = select.price > 100
    print("   select.price > 100")
    print(f"   → {query}\n")

    print("2. Range query:")
    query = select.price.is_between(start=50, end=500)
    print("   select.price.is_between(start=50, end=500)")
    print(f"   → {query}\n")

    print("3. IN query:")
    query = select.category.is_in(items=["electronics", "furniture"])
    print("   select.category.is_in(items=['electronics', 'furniture'])")
    print(f"   → {query}\n")

    print("4. Boolean query:")
    query = select.in_stock.is_true()
    print("   select.in_stock.is_true()")
    print(f"   → {query}\n")

    print("5. Logical AND (match):")
    query = select.match(select.price > 50, select.price < 500, select.in_stock.is_true())
    print("   select.match(")
    print("       select.price > 50,")
    print("       select.price < 500,")
    print("       select.in_stock.is_true()")
    print("   )")
    print(f"   → {query}\n")

    print("6. Logical OR (at_least):")
    query = select.at_least(select.price < 50, select.price > 1000)
    print("   select.at_least(")
    print("       select.price < 50,")
    print("       select.price > 1000")
    print("   )")
    print(f"   → {query}\n")

    print("7. Sorting:")
    asc_sort = select.price.asc()
    desc_sort = select.name.desc()
    print(f"   select.price.asc() → {asc_sort}")
    print(f"   select.name.desc() → {desc_sort}\n")

    print("8. Complex nested query:")
    query = select.at_least(select.match(select.category == "electronics", select.price >= 100), select.match(select.category == "furniture", select.in_stock.is_true()))
    print("   select.at_least(")
    print("       select.match(")
    print("           select.category == 'electronics',")
    print("           select.price >= 100")
    print("       ),")
    print("       select.match(")
    print("           select.category == 'furniture',")
    print("           select.in_stock.is_true()")
    print("       )")
    print("   )")
    print(f"   → {query}\n")

    print("✅ ORM-style query examples completed!")

File: examples-js/query_operators_example.js
const { MongoClient } = require('mongodb');
const { Schema, Flexmodel, field } = require('../dist/index');

class Product extends Flexmodel {
  static schema = Schema.ident({
    name: field(String, { nullable: false }),
    price: field(Number, { default: 0.0 }),
    inStock: field(Boolean, { default: true }),
  });

  constructor(data = {}) {
    super(data);
    if (!this.name) {
      this.name = Product.schema.get('name').default;
    }
    if (this.price === undefined) {
      this.price = Product.schema.get('price').default;
    }
    if (this.inStock === undefined) {
      this.inStock = Product.schema.get('inStock').default;
    }
  }
}

async function main() {
  try {
    const client = new MongoClient('mongodb://localhost:27017/testdb', {
      serverSelectionTimeoutMS: 1000,
    });
    
    await client.connect();
    
    Product.attach(client.db(), 'products');

    // Create a select query using Proxy-based field access
    const select = Product.select();

    // Note: In JavaScript/TypeScript, we can't override comparison operators (==, <, >, etc.)
    // So we need to use method calls instead
    select.where(select.name.eq('Shoes'));
    select.where(
      select.match(
        select.atLeast(
          select.inStock.isFalse(),
          select.inStock.isTrue()
        ),
        select.price.lt(100)
      )
    );

    console.log('Query built successfully!');
    console.log(`Query: ${select.queryString}`);
    console.log(`SQL: ${select.toSql}`);

    // Try to fetch results (will fail if MongoDB is not running)
    try {
      const results = await select.fetchAll();
      for (const item of results) {
        console.log(`Product: ${item.name}, Price: ${item.price}, In Stock: ${item.inStock}`);
      }
    } catch (e) {
      console.log(`\n⚠️  Could not fetch from database: ${e.message}`);
      console.log('This is expected if MongoDB is not running.');
    }

    await client.close();
  } catch (e) {
    console.log(`⚠️  MongoDB not available: ${e.message}`);
    console.log('To run this example with database access, start MongoDB at localhost:27017');
  }
}

main().catch(console.error);
